docs: add EPP flow control reference (#1130)
llm-d/llm-d: docs/wip-docs-new/architecture/core/epp/configuration.md

Reviewer: try to be more concise here; "The `saturationDetector` section configures the saturation detection mechanism" doesn't offer information

Author: Done!

> - The [`saturationDetector`](#saturationDetector) section configures the saturation detector.

Reviewer: We should consolidate saturationDetector under flowControl, one section in the config file per component.

Author: I've consolidated them conceptually by placing `saturationDetector` under the Flow Control heading in the document. The YAML examples still show it at the top level to remain accurate to the current API and avoid implying a breaking schema change.

I will update this in the PR that code change takes effect.

> #### `flowControl`

Reviewer: We have too much details about flow control in this configuration-focused file; I think we should move most of this to the flow-control.md file, and here we only reference and focus only on describing the flow control related fields in the config. It should be concise and to the point.

Author: Done!

The `flowControl` section configures the EPP's Flow Control layer, which acts as a pool defense mechanism by buffering requests before they reach backend model servers. Flow Control implements a 3-tier dispatch hierarchy: **Priority → Fairness → Ordering**. For a visual breakdown of how this looks in practice, see the [Queuing Topology diagram in the Flow Control reference](flow-control.md#queuing-topology--the-3-tier-dispatch).

When flow control is enabled (via the `FlowControl` feature gate), incoming requests are queued in memory and dispatched according to configured priority bands, fairness policies, and ordering policies. When the pool is saturated (as determined by the [saturation detector](#saturationdetector)), requests are held in the queue until capacity frees up.

> The following example demonstrates how to supply this configuration via the Helm chart's `values.yaml` file, using the `pluginsCustomConfig` feature to inject the `EndpointPickerConfig`. It shows how to configure the `featureGates`, the `plugins` list, the `saturationDetector`, and the `flowControl` section to work together.

Reviewer: Are we planning to always have flow control guarded by a flag? if yes, we should move the feature gate to become an explicit flag under the config. Feature gates are temporary.

Author: No, the plan is to have it _always on_ in the subsequent release (pending final scale testing and user feedback). Flow Control defaults work well for single-tenant use cases. There is no real reason to toggle-off.

Author: For now, I need to document the feature gate for enablement.

```yaml
inferenceExtension:
  pluginsConfigFile: "custom-plugins.yaml"
  pluginsCustomConfig:

>     custom-plugins.yaml: |

Reviewer: pls remove the above lines, those are part of the helm chart; lets focus on the config api only here.

Author: Done!

      apiVersion: inference.networking.x-k8s.io/v1alpha1
      kind: EndpointPickerConfig

      featureGates:
      - flowControl

      plugins:

>       - name: round-robin-fairness-policy

Reviewer: no need to explicitly name the plugin, it will default to the type. this will make the config more concise

Author: Done!

        type: round-robin-fairness-policy
      - name: fcfs-ordering-policy
        type: fcfs-ordering-policy
      - name: global-strict-fairness-policy
        type: global-strict-fairness-policy
      - name: my-concurrency-detector
        type: concurrency-detector
      # ... other plugins ...

      saturationDetector:
        pluginRef: my-concurrency-detector

>       flowControl:

Reviewer: If that is not the case already, I suggest to put the present the configuration using the default values and clearly indicate that those are defaults.

Author: Thanks for the suggestion! I have updated the `flowControl` section in to use the actual default values in the YAML example and added `# Default` comments to make it explicit. Let me know if you think it needs to be even clearer or applied differently.

Author: There are 0 required flow control fields outside of the feature gate. Everything has practical, cascading defaults if unspecified at any level.

> ##### Fairness Policies

Reviewer: The list of available plugins should be maintained in one place, and I suggest to have them in the flow-control.md file; here we only describe the parameters that one can use to select the various pluggable behaviors, and we reference flow-control.md for the reader to lookup available options.

Author: Removed (as I already had a list in flow-control.md). I also enhanced the descriptions in that file slightly to serve as a cheat-sheet while still remaining concise.

Fairness policies control how requests from different flows (e.g., different tenants) are interleaved within a priority band.

*   **[`global-strict-fairness-policy`](placeholder)**: Ignores flow isolation and serves all requests in a single global order based on the Ordering Policy.
*   **[`round-robin-fairness-policy`](placeholder)**: Guarantees fair sharing by cycling through active flows one by one.

##### Ordering Policies

Ordering policies control the order in which requests are dispatched from the queue within a given flow.

*   **[`fcfs-ordering-policy`](placeholder)**: First-Come, First-Served based on arrival time. (Default)
*   **[`edf-ordering-policy`](placeholder)**: Earliest Deadline First, prioritizing requests with the closest expiration time.
*   **[`slo-deadline-ordering-policy`](placeholder)**: Orders requests by an SLO-based deadline computed from arrival time.

> See [Flow Control](flow-control.md) for more architectural details on how the EPP's flow control layer works internally.

Reviewer: this sentence should be put at the top of the flow control section

Author: Moved.

The behavior of the saturation detector depends on whether flow control is enabled:

- **Flow Control enabled**: When the pool is saturated, request dispatch is paused and incoming requests are buffered in the flow control memory queues (respecting priority and fairness policies) until backend capacity frees up. The Saturation Detector acts as the gatekeeper for these centralized queues; see the [Dispatch Lifecycle section in the Flow Control reference](flow-control.md#the-dispatch-lifecycle) for details.

> - **Flow Control disabled** (default): When the pool is saturated, "sheddable" requests (those with negative priority) are immediately rejected with HTTP 429 (Too Many Requests). All other requests pass directly to the model servers.

Reviewer: all the above is better discussed in the flow-control section, here only discuss the knobs available for configuring saturation detection.

Author: Moved the behavior toggling description to the other doc and streamlined this section of the config guide.

##### Fields

> - `pluginRef`: References a plugin instance defined in the global `plugins` section. Defaults to `utilization-detector` if omitted or empty. *Note: If a `utilization-detector` is not explicitly defined in your `plugins` array, the gateway will automatically instantiate one under the hood using standard default parameters.*

Reviewer: ```suggestion
- `pluginRef`: References a plugin instance defined in the global `plugins` section. Defaults to `utilization-detector` if omitted or empty. *Note: If a `utilization-detector` is not explicitly defined in the `plugins` array, one will automatically be instantiated using default parameters.*
```

Author: Applied with minor tweaks.

> [!NOTE]
> #### The "Healthy Buffer" Principle

> > Regardless of which detector you use, the core goal of tuning saturation detection is to maintain a small, **"healthy buffer"** of requests queued locally on the model servers themselves. This buffer should be just large enough to ensure continuous batching engines never starve for work, but small enough that the vast majority of queuing happens centrally in the EPP where priority and fairness can be enforced.

Reviewer: It is not clear from this paragraph what knob a user should use to do this. 

Also, this paragraph seems to be a better fit under flow-control.md.

> ##### Saturation Detector Plugins

Reviewer: ditto, list of plugins should be in one place, under flow-control.md

#### Flow Control Metrics

> When flow control is enabled, the system exposes detailed metrics for queue depths, dispatch cycles, and saturation. For a full list of these metrics and their labels, see the [Flow Control reference](flow-control.md#metrics--observability).

Reviewer: ditto, this has nothing to do with configuration, better fit in flow-control.md

Author: Removed.
